rosbag2_storage/src/rosbag2_storage/metadata_io.cpp

```diff
@@ -39,6 +39,24 @@
 
 namespace YAML
 {
+/// Allow pass a metadata version to the sub-structs of BagMetadata for deserializing.
+/**
+  * Encoding should always use the current metadata version, so it does not need this value.
+  * We cannot extend the YAML::Node class to include this, so we must call it
+  * as a function with the node as an argument.
+  */
+template<typename T>
+T versioned_decode(const Node & node, int version)
+{
```

Comment:
Consider adding a static_assert to check that T is "correct".

Reply:
I don't understand, where would this assert go?

Reply:
this assert could be the first line of the function. Alternatively it could go in the template declaration with `std::enable_if`, but I don't think that's needed.
Can we limit the amount of `T`s for this function?

Reply:
Sorry, could one of you use the github suggestion feature or a code block, I don't understand what you're looking to accomplish

Reply:
So further below, we call this function with `T = rosbag2_storage::TopicInformation`. For this type, the function is well defined.
```
decode_for_version<std::vector<rosbag2_storage::TopicInformation>>
```

With  a `static_assert` we could verify that this function is not called with a type T which doesn't behave correctly. One which comes to my mind here is `std::is_trivially_default_constructible<T>` as we at one point call `T value`. The first line could then be something like:
```
static_assert(std::is_trivially_default_constructible<T>::value);
```
That would prevent a compilation with a wrong type T.

That is all to say, that it's most likely not being needed given that all of this is in a `.cpp` file in the first place and can't be called by anybody else than this file.

Reply:
I have added an assertion for `std::is_default_constructible`, the `trivially` version doesn't work here.